### PR TITLE
Allow reverse iteration order during resource permissions checking

### PR DESCRIPTION
**Pull request by jmckenzie-dev**
Patch by Sam Tunnicliffe; reviewed by Blake Eggleston, Jeff Jirsa, and ???? for CASSANDRA-17016

Co-authored by Sam Tunnicliffe <sam@beobal.com>
Co-authored by Josh McKenzie <jmckenzie@apache.org>

### CODE REVIEW

File: CHANGES.txt.orig [hunk 1]
```diff
@@ -0,0 +1,7203 @@
+4.1
```
Comment (azotcsit) [inline CHANGES.txt.orig]: This file is smth related to merge process, it needs to be deleted

File: test/unit/org/apache/cassandra/auth/AuthTestUtils.java [hunk 2]
```diff
@@ -18,7 +18,12 @@
 
 package org.apache.cassandra.auth;
 
+import java.util.Map;
```
Comment (azotcsit) [inline test/unit/org/apache/cassandra/auth/AuthTestUtils.java]: Are these imports required?

File: src/java/org/apache/cassandra/service/ClientState.java [hunk 3]
```diff
@@ -425,9 +427,22 @@ public void ensurePermission(Permission permission, Function function)
 
     private void ensurePermissionOnResourceChain(Permission perm, IResource resource)
     {
-        for (IResource r : Resources.chain(resource))
-            if (authorize(r).contains(perm))
-                return;
+        if (DatabaseDescriptor.getAuthFromRoot())
```
Comment (azotcsit) [inline src/java/org/apache/cassandra/service/ClientState.java:432]: Even though the current implementation looks good, I feel there is a small point for improvement:
```
        List<? extends IResource> resources = Resources.chain(resource);
        if (DatabaseDescriptor.getAuthFromRoot())
            resources = Lists.reverse(resources);

        for (IResource r : resources)
            if (authorize(r).contains(perm))
                return;
```

As for me this version has a few minor advantages:
1. more compact
2. does not have cast operation (I generally consider them unsafe)
3. does not have duplicated "looping logic"

File: CHANGES.txt.orig (hunk 1 above)
Comment (jmckenzie-dev) [reply CHANGES.txt.orig]: My git's been greedy adding things; will fix.

File: test/unit/org/apache/cassandra/auth/AuthTestUtils.java (hunk 2 above)
Comment (jmckenzie-dev) [reply test/unit/org/apache/cassandra/auth/AuthTestUtils.java]: Removed the various unused imports there.

File: src/java/org/apache/cassandra/service/ClientState.java (hunk 3 above)
Comment (jmckenzie-dev) [reply src/java/org/apache/cassandra/service/ClientState.java:432]: The casting we can fix w/the current implementation if we're so inclined but I agree that the compactness and readability are improved with this revision.

A minor observation: I don't know if a reverse iterator from the java libs or Lists.reverse's held state from guava are comparable or not in terms of state, but whatever differences there may be there I expect to be trivial enough to be more than made up for in code clarity and lack of repetition.

**Comment by jmckenzie-dev** [comment]
Updated and a new commit pushed w/the changes. Thanks for the feedback.

**Review by azotcsit — APPROVED** [review]
LGTM

**Pull request closed without merging**